src/typescript-reporter/reporter/TypeScriptReporter.ts

```diff
@@ -0,0 +1,171 @@
+import * as ts from 'typescript';
+import { FilesChange, Reporter } from '../../reporter';
+import { createIssuesFromTsDiagnostics } from '../issue/TypeScriptIssueFactory';
+import { TypeScriptReporterConfiguration } from '../TypeScriptReporterConfiguration';
+import { createControlledWatchCompilerHost } from './ControlledWatchCompilerHost';
+import { TypeScriptExtension } from '../extension/TypeScriptExtension';
+import { createTypeScriptVueExtension } from '../extension/vue/TypeScriptVueExtension';
+import { createTypeScriptPnpExtension } from '../extension/pnp/TypeScriptPnpExtension';
+import { createControlledWatchSolutionBuilderHost } from './ControlledWatchSolutionBuilderHost';
+import {
+  ControlledTypeScriptSystem,
+  createControlledTypeScriptSystem,
+} from './ControlledTypeScriptSystem';
+
+function createTypeScriptReporter(configuration: TypeScriptReporterConfiguration): Reporter {
+  const extensions: TypeScriptExtension[] = [];
+
+  let system: ControlledTypeScriptSystem | undefined;
+  let watchCompilerHost:
+    | ts.WatchCompilerHostOfConfigFile<ts.SemanticDiagnosticsBuilderProgram>
+    | undefined;
+  let watchSolutionBuilderHost:
+    | ts.SolutionBuilderWithWatchHost<ts.SemanticDiagnosticsBuilderProgram>
+    | undefined;
+  let watchProgram: ts.WatchOfConfigFile<ts.SemanticDiagnosticsBuilderProgram> | undefined;
+  let solutionBuilder: ts.SolutionBuilder<ts.SemanticDiagnosticsBuilderProgram> | undefined;
+
+  const diagnosticsPerProject = new Map<string, ts.Diagnostic[]>();
+
+  if (configuration.extensions.vue.enabled) {
+    extensions.push(createTypeScriptVueExtension(configuration.extensions.vue));
+  }
+  if (configuration.extensions.pnp.enabled) {
+    extensions.push(createTypeScriptPnpExtension());
+  }
```

Comment:
Could this be a good moment to move those extensions out to separate packages and make them injectable via configuration?

Reply:
I would keep them in the same package for now. In my opinion, these extensions are a workaround and I hope that the TypeScript core team will agree to add API for compiler plugins. Currently we need to implement these extensions in the plugin, loader, and language service. I've created https://github.com/microsoft/TypeScript/issues/38736 proposal based on the experience with the plugin - maybe it will move forward. If they will not agree, then we can think about a unified way to write TypeScript plugins that could work both in the plugin and ts-loader :)

Reply:
I'm rooting for it. Let's see how it goes :)